lib/rgeo/cartesian/calculations.rb

```diff
@@ -69,47 +69,236 @@ def contains_point?(p)
         sx2 = s2.x
         sy2 = s2.y
         dx2 = seg.dx
         dy2 = seg.dy
         denom = @dx * dy2 - @dy * dx2
+
         if denom == 0
           # Segments are parallel. Make sure they are collinear.
-          return false unless side(s2) == 0
-          # 1-D check.
-          ts = (@dx * (sx2 - @sx) + @dy * (sy2 - @sy)) / @lensq
-          te = (@dx * (sx2 + dx2 - @sx) + @dy * (sy2 + dy2 - @sy)) / @lensq
-          if ts < te
-            te >= 0.0 && ts <= 1.0
-          else
-            ts >= 0.0 && te <= 1.0
-          end
+          return nil unless side(s2) == 0
+
+          # return the first point it finds that intersects another line.
+          # In many cases, the intersection is actually another line
+          # segment, but for now, we will just return a single point.
+          return s2 if contains_point?(s2)
+          return seg.e if contains_point?(seg.e)
+          return @s if seg.contains_point?(@s)
+          return @e if seg.contains_point?(@e)
+          nil
         else
           # Segments are not parallel. Check the intersection of their
           # containing lines.
-          t = (dy2 * (sx2 - @sx) + dx2 * (@sy - sy2)) / denom
-          return false if t < 0.0 || t > 1.0
-          t2 = (@dy * (sx2 - @sx) + @dx * (@sy - sy2)) / denom
-          t2 >= 0.0 && t2 <= 1.0
+          num1 = dx2 * (@sy - sy2) - (dy2 * (@sx - sx2))
+          num2 = @dx * (@sy - sy2) - (@dy * (@sx - sx2))
+          cross1 = num1 / denom
+          cross2 = num2 / denom
+
+          return nil if cross1 < 0.0 || cross1 > 1.0
+          if cross2 >= 0.0 && cross2 <= 1.0
+            x = @sx + (cross1 * @dx)
+            y = @sy + (cross1 * @dy)
+
+            # Check if this segment contains the point.
+            # Sometimes round-off errors occur and intersections
+            # are recorded as off the line segments.
+            #
+            # If this is the case, return the closest point from
+            # either segment.
+            int_pt = @s.factory.point(x, y)
+            if contains_point?(int_pt)
+              int_pt
+            else
+              # find closest of @s, @e, seg.s, seg.e
+              [@e, seg.s, seg.e].reduce(@s) do |closest, pt|
+                int_pt.distance(pt) < int_pt.distance(closest) ? pt : closest
+              end
+            end
+          end
         end
       end
 
       def length
         Math.sqrt(@lensq)
       end
     end
+
+    # Implements a Sweepline intersector to find all intersections
+    # in a group of segments. The idea is to use a horizontal line starting
+    # at y = +Infinity that sweeps down to y = -Infinity and every time it hits
+    # a new line, it will check if it intersects with any of the segments
+    # the line currently intersects at that y value.
+    # This is a more simplistic implementation that uses an array to hold
+    # observed segments instead of a sorted BST, so performance may be significantly
```

Comment:
I am a bit unsure of where it would be more appropriate, could you indicate me ?

Reply:
![image](https://user-images.githubusercontent.com/12189611/124161023-e9de9880-da6a-11eb-8f13-f8d7e6da5138.png)
```
LINESTRING(1 1, 0 1, 0 0, 0.125 0.5, 0.25 0, 0.375 0.5, 0.5 0, 0.625 0.5, 0.75 0, 0.875 0.5, 1 0, 1 1)
```

In this case, since we are comparing all `observed_segments` against each other, this would be about O(n^2) since pretty much every segment is being observed at once here. If we were using a BST instead of an array or set, we would sort it so that only adjacent segments are compared to each other, but it's a lot more work to implement the tree and keep track of switches in order. 

In this case,

![image](https://user-images.githubusercontent.com/12189611/124161554-8bfe8080-da6b-11eb-95a6-7f54e51d73a6.png)

```
LINESTRING (2.0 0.0, 1.847759065 -0.7653668647, 1.4142135624 -1.4142135624, 0.7653668647 -1.847759065, 0.0 -2.0, -0.7653668647 -1.847759065, -1.4142135624 -1.4142135624, -1.847759065 -0.7653668647, -2.0 0.0, -1.847759065 0.7653668647, -1.4142135624 1.4142135624, -0.7653668647 1.847759065, 0.0 2.0, 0.7653668647 1.847759065, 1.4142135624 1.4142135624, 1.847759065 0.7653668647, 2.0 0.0)
```

There are only ever 2 `observed_segments` at once so it will run closer to O(nlogn).

Reply:
There's also another strategy that GEOS uses where they break the segments up into monotone pieces and run the intersection on each piece then combine the results. 

I don't think it's really necessary we do either of these. If performance is that big of a concern you should use GEOS instead of the simpler Ruby implementation.

Reply:
@keithdoggett I definitely agree, thanks for the pointer though!